python/pyspark/sql/dataframe.py

```diff
@@ -1761,12 +1761,37 @@ def toPandas(self):
                 raise ImportError("%s\n%s" % (e.message, msg))
         else:
             dtype = {}
+            columns_with_null_int = {}
+            def null_handler(rows, columns_with_null_int):
+                for row in rows:
+                    row = row.asDict()
+                    for column in columns_with_null_int:
+                        val = row[column]
+                        dt = dtype[column]
+                        if val is not None:
```

Comment:
Don't we want to change data type for `None` values? I don't see you do it?

Reply:
They are handled by Pandas already, so I am just letting them pass through.

Reply:
Don't we want to fix the issue when pandas type in (np.int8, np.int16, np.int32) and the field is nullable, the `dtype` we get will cause exception later when converting a `None` to int type such as np.int16?

Reply:
You can follow the https://github.com/apache/spark/pull/18945/files#r134033952 suggested.

Reply:
If `pandas_type in (np.int8, np.int16, np.int32) and field.nullable` and there are ANY non-null values, the dtype of the column is changed to `np.float32` or `np.float64`, both of which properly handle `None` values.

That said, if the entire column was `None`, it would fail. Therefore I have preemptively changed the type on line 1787 to `np.float32`. Per `null_handler`, it may still change to `np.float64` if needed.

Reply:
If `pandas_type in (np.int8, np.int16, np.int32) and field.nullable` and there are ANY non-null values, the dtype of the column is changed to `np.float32` or `np.float64`, both of which properly handle `None` values.

That said, if the entire column was `None`, it would fail. Therefore I have preemptively changed the type on line 1787 to `np.float32`. Per `null_handler`, it may still change to `np.float64` if needed.